### ISSUE TITLE
Lots of stored references to player objectrefs

### ISSUE THREAD

**Issue by Lazerbeak12345** [blert2112/lib_mount#5]
[This section of the official lua modding documentation](https://github.com/minetest/minetest/blob/master/doc/lua_api.md#advice-on-handling-objectrefs) does not recommend keeping pointers to object refs (I understand a player to be an example of one - perhaps I'm wrong). This is to avoid the objectref going invalid.

Looking over this code, pretty much every single reference to a "player" is not the player's name - but an objectref. This is only a problem where the objectref (and not the username) is stored in the vehicle (for the driver and all passengers), and the potentially completely useless `lib_mount.passengers` value (which is even more scary - you're indexing with a playerref to get the same playerref, and the same playerref might not even == itself all the time)

This is brought up as an issue instead of a PR because fixing this ~~_will_~~ may break backwards compatibility.

---

See below. There's a way, at cost of performance, to not break compatibility.

**Comment by Lazerbeak12345** [comment]
One example of where this is _not_ a problem: It's okay for `force_detach` to take a playerref.

It's _not_ okay for `entity.driver` to be a playerref (as a user could, for example, ride around their vehicle on a public server for a bit, go afk for a very long time, then come back to find that they are stuck to a car that they can't control, and can't unmount from (their attempts to do so may even crash the server)

**Comment by Lazerbeak12345** [comment]
Perhaps related to #3

**Comment by Panquesito7** [comment]
Would there be any way to keep backwards compatibility? We could check the data of the entity to see if it has old data. Or just provide an option in the `settingtypes.txt` file that when enabled it'll use the new code (with a note, of course).

**Comment by Lazerbeak12345** [comment]
It's been awhile since I looked at this, but a non-breaking way might be to store both playerref and playername? There'd have to be some way to detect the objectref is invalid, and if it is, update it to match the value represented by a string.

As for the table thing, it could be `(playername : Boolean) ∪ (playerref : playerref)`, keeping compatibility, but still functioning when refs are invalid. It'd still collect garbage, and is still a memory leak - if a slow one. A slower approach is to use metatables, and if it's indexed with an object ref then it gets the name, checks if it's in the real table, then returns nil or the entered ref accordingly. Perhaps the real table should be exposed for performance reasons?

In all, I think I was mistaken. It's fixable.

---

Most of these assumptions were incorrect. See below.

**Comment by Lazerbeak12345** [comment]
The real question is this: which should be the source of truth? The old API or the new? What does it mean if the playerref and playername are both valid, but they are different players?

---

I didn't realize this at the time, but if the old API were the source of truth, it would accomplish nothing, as it would still store player refs.

**Comment by Lazerbeak12345** [comment]
Arguably, that state should be avoided. The new API should make the question irrelevant.

**Comment by Lazerbeak12345** [comment]
Ah. That's the problem. They intentionally _don't_ provide a way to tell if it's a valid Playerref: 

> However, doing this is NOT recommended as there is (intentionally) no method to test if a previously acquired ObjectRef is still valid. Instead, ObjectRefs should be "let go" of as soon as control is returned from Lua back to the engine. Doing so is much less error-prone and you will never need to wonder if the object you are working with still exists.

**Comment by Lazerbeak12345** [comment]
So it looks like, with that in mind, here's the rough idea of the change that needs to be made:

- For tables that store only playerrefs, make a new table that is named to be clear that it's storing player names, then re-implement the old table using metatables. New is the source of truth.
- Functional apis accepting player info will now also accept playernames instead of player refs.

The real challenge is the data-oriented apis.

Here's the table of which API is using which type of player info:

| \                      | Stores ref | Accepts ref |
|:-----------------------|------------|-------------|
| `entity.passengers[x].player`    | x          |             |
| `entity.driver`        | x          |             |
| `lib_mount.attach`      |            | x           |
| `lib_mount.detach`      |            | x           |
| `lib_mount.passengers` | x          |             |

Clearly, for `attach` and `detach`, we can leave it as-is.

- `lib_mount.passengers` can become a metatable, updating the state of a new table, `lib_mount.passengers_by_name` (string : true), and storing nothing. It can also get its references from that other tree.
- `entity.driver` and `entity.passengers[x].player` is hard. Options I have in mind, with problems listed:
    - Make the whole table a metatable
      - Problem: Very slow
    - Store both (impossible)
      - Problem: The ref would need to support updates, but how would you know if it was the name updated, or if it was the ref updated?
      - Problem: The ref goes invalid eventually. You can't tell it's invalid, so you are forced to ignore the ref in most cases. (Breaking change)
    - Store only the name
      - Problem: This is a breaking change.

**Comment by Lazerbeak12345** [comment]
TLDR: Sorry for all the pings, and yes, with research, it's possible to not have any breaking changes, at the cost of performance.

**Comment by Panquesito7** [comment]
> TLDR: Sorry for all the pings, and yes, with research, it's possible to not have any breaking changes, at the cost of performance.

I'd say we should make this change, maybe add an important note in the release and the readme.
We can detect if the user is still using the old API, and if so, send a message to the administrators in-game, or simply not let the server to run by sending an error message.

What do you think? And thank you for your detailed explanation!

**Comment by Lazerbeak12345** [comment]
If we're making those sort of logs for the admin to see, we should also log the other API change we made a few years ago.

**Comment by Panquesito7** [comment]
> If we're making those sort of logs for the admin to see, we should also log the other API change we made a few years ago.

That's right, although that should be a warning as IIRC it didn't bring breaking changes.
However, if we plan on performing these changes, we should probably send an error (crash the server) when joining if it's still using the old API, if we can somehow detect that.

**Comment by Lazerbeak12345** [comment]
So is the approach you want the "Store only the name" approach from my research? It's the one with the least breaking changes, though the one change that is broken can't easily be detected. The only way I can think of is to make a "mock" playerref, with a metatable setup so if any indexing happens, it asserts. A problem is that it's very easy to detect the difference between the real and fake, because `type` returns `userdata` for the real and `table` for the fake. Can be an issue since some mods rely on the `type` being correct. We want to assert before they do.

**Comment by Lazerbeak12345** [comment]
3d armor seems to handle fake playerrefs gracefully. I don't think we need to worry about my type concerns.

**Comment by Panquesito7** [comment]
Thank you for your research and your help. What needs to be done to fix this issue properly, then?

**Comment by Lazerbeak12345** [comment]
Looks like the decision comes down to one of these

- Use potentially slow metatables to retain backwards compatibility (may be slow even with new API)
- Break compatibility
- No change is also an option, but it's unclear if it will cause problems, since the doc advises against it.

**Comment by Panquesito7** [comment]
I'd suggest we keep backwards compatibility for some time to warn users about the change, and then, fully break compatibility in a later version.
